HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,440 @@
+  desc.add<double>("MinDXYBeamSpotL3", -1.0);
+  desc.add<double>("MaxDzL3", 9999.0);
+  desc.add<double>("MinDxySigL3", -1.0);
+  desc.add<double>("NSigmaPtL3", 0.0);
+  desc.add<double>("MaxPtDifferenceL3", 9999.0);
+  desc.add<double>("MinTrackPtL3", 0.0);
+  desc.add<double>("MaxNormalizedChi2L3", 9999.0);
+  desc.add<double>("MinPtL3", 23.0);
+  desc.add<double>("MaxEtaL3", 2.5);
+  desc.add<int>("MinNmuonHitsL3", 0);
+  desc.add<std::vector<int>>("MinNhitsL3", {1, 0});
+  desc.add<std::vector<double>>("AbsEtaBinsL3", {1, 9999.});
+  desc.add<std::vector<int>>("MinNstationsL3", {1, 1});
+  desc.add<std::vector<int>>("MinNchambersL3", {1, 0});
+  desc.add<int>("MinNL2", 0);
+  desc.add<double>("MaxDXYBeamSpotL2", 9999.0);
+  desc.add<double>("MinDXYBeamSpotL2", -1.0);
+  desc.add<double>("MaxDzL2", 9999.0);
+  desc.add<double>("MinDxySigL2", -1.0);
+  desc.add<double>("NSigmaPtL2", 0.0);
+  desc.add<double>("MaxPtDifferenceL2", 9999.0);
+  desc.add<double>("MinTrackPtL2", 0.0);
+  desc.add<double>("MaxNormalizedChi2L2", 9999.0);
+  desc.add<double>("MinPtL2", 23.0);
+  desc.add<double>("MaxEtaL2", 2.5);
+  desc.add<int>("MinNmuonHitsL2", 0);
+  desc.add<std::vector<int>>("MinNhitsL2", {1, 0});
+  desc.add<std::vector<double>>("AbsEtaBinsL2", {1, 9999.});
+  desc.add<std::vector<int>>("MinNstationsL2", {1, 1});
+  desc.add<std::vector<int>>("MinNchambersL2", {1, 0});
+  desc.add<bool>("MatchToPreviousCand", true);
+  desc.add<double>("MaxDeltaR2L2L3", 0.0025);
+  descriptions.add("hltMuonL3orL2PreFilter", desc);
+}
+
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags()) {
+    filterproduct.addCollectionTag(L3candTag_);
+    filterproduct.addCollectionTag(L2CandTag_);
+  }
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3candToken_);
+  auto const L2mucands = iEvent.getHandle(L2CandToken_);
+  auto const previousLevelCands = iEvent.getHandle(previousCandToken_);
+  std::vector<reco::RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(trigger::TriggerMuon, vcands);
+  }
+
+  auto const& beamSpot = iEvent.get(beamspotToken_);
+
+  // Number of objects passing the L3 Trigger:
+  int nL3 = 0;
+  for (unsigned int iMuL3 = 0; iMuL3 < L3mucands->size(); iMuL3++) {
+    reco::RecoChargedCandidateRef L3cand(L3mucands, iMuL3);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L3cand isNonnull " << L3cand.isNonnull();
+    // safeguard
+    if (L3cand->pt() == 0.0) {
+      LogDebug("HLTMuonL3orL2PreFilter") << "Skipping candidate: L3cand pT is 0";
+      continue;
+    }
```

Comment:
```suggestion

```

Reply:
Do I have to remove this? In a previous comment you suggested to add this to avoid divisions by 0.